google-cloud-pubsub/lib/google/cloud/pubsub/flow_controller.rb

```diff
@@ -0,0 +1,141 @@
+          unless [:ignore, :error, :block].include? limit_exceeded_behavior
+            raise ArgumentError, "limit_exceeded_behavior must be one of :ignore, :error, :block"
+          end
+          if [:error, :block].include?(limit_exceeded_behavior) && message_limit < 1
+            raise ArgumentError,
+                  "Flow control message limit (#{message_limit}) exceeded by a single message, would block forever"
+          end
+          @mutex = Mutex.new
+          @message_limit = message_limit
+          @byte_limit = byte_limit
+          @limit_exceeded_behavior = limit_exceeded_behavior
+          @outstanding_messages = 0
+          @outstanding_bytes = 0
+
+          @awaiting = []
+        end
+
+        def acquire message_size
+          return if limit_exceeded_behavior == :ignore
+          @mutex.lock
+          if limit_exceeded_behavior == :error && would_exceed_message_limit
+            raise FlowControlLimitError, "Flow control message limit (#{message_limit}) would be exceeded"
+          end
+          if limit_exceeded_behavior == :error && would_exceed_byte_limit(message_size)
+            raise FlowControlLimitError,
+                  "Flow control byte limit (#{byte_limit}) would be exceeded, message_size: #{message_size}"
+          end
+          if limit_exceeded_behavior == :block && message_size > byte_limit
+            raise FlowControlLimitError,
+                  "Flow control byte limit (#{byte_limit}) exceeded by a single message, would block forever"
+          end
+
+          acquire_or_wait message_size
+        ensure
+          @mutex.unlock if @mutex.owned?
+        end
+
+        def release message_size
+          return if limit_exceeded_behavior == :ignore
+          @mutex.synchronize do
+            # Releasing a message decreases the load.
+            @outstanding_messages -= 1
+            @outstanding_bytes -= message_size
+            if @outstanding_messages.negative? || @outstanding_bytes.negative?
+              # Releasing a message that was never added or already released.
+              @outstanding_messages = [0, @outstanding_messages].max
+              @outstanding_bytes = [0, @outstanding_bytes].max
+            end
+            @awaiting.first.set unless @awaiting.empty?
+          end
+        end
+
+        protected
+
+        # rubocop:disable Style/IdenticalConditionalBranches
+        # rubocop:disable Style/GuardClause
+
+        def acquire_or_wait bytes_remaining
+          waiter = nil
+          while would_exceed_byte_limit(bytes_remaining) || would_exceed_message_limit
+            if would_exceed_byte_limit bytes_remaining
+              # Take what is available, even if still blocked by message limit as well as byte limit.
+              available = byte_limit - @outstanding_bytes
+              bytes_remaining -= available
+              @outstanding_bytes = byte_limit
+            end
```

Comment:
Now that @dazuma mentioned it (good catch, BTW), I took another look, and this is indeed the thing where Ruby implementation differs from Python's.

The problem is that there we only have common pool of available bytes (`@outstanding_bytes`) freely accessible to all threads. A thread that arrives later can thus steal some chunk of the available bytes capacity even if there are already other threads waiting for the same. And then we can end up with multiple threads blocked, because each only partially reserved the needed capacity, and is now waiting for more even though all partial reservations combined would be enough to release on the threads..

In Python (and Java, IIRC) this is solved by having a [FIFO queue](https://github.com/googleapis/python-pubsub/blob/de5429aa1663c634705e2c4abf62d7aa7c375316/google/cloud/pubsub_v1/publisher/flow_controller.py#L51-L58) tracking the order of the threads and how many bytes each thread has reserved so far. When a message is released, the freed up capacity is first [distributed](https://github.com/googleapis/python-pubsub/blob/de5429aa1663c634705e2c4abf62d7aa7c375316/google/cloud/pubsub_v1/publisher/flow_controller.py#L183) among the waiting threads in FIFO order before waking them up.

---

To a lesser extent, this also applies to the _message count_ limit. When a message is released and a slot opens up to accept one new message, that slot might get stolen by a new thread, because it might just sneak in before the old thread currently waiting can re-acquire the lock and proceed.

That won't cause a deadlock, because you can't reserve "half a message", and it probably won't cause starving, because eventually the old waiting thread gets lucky and acquires the lock first, but it would still be nice if a publish request is first sent for earlier messages.

(discovered that in Python after re-visiting the code).

Reply:
@dazuma I converted your scenario above into a unit test:

```rb
it "deadlocks" do
      100.times do |i|
        flow_controller = Google::Cloud::PubSub::FlowController.new(
          message_limit: 1000,
          byte_limit: 100,
          limit_exceeded_behavior: :block
        )

        adding_1_done = Concurrent::Event.new
        adding_2_done = Concurrent::Event.new
        adding_3_done = Concurrent::Event.new

        _(flow_controller.outstanding_bytes).must_equal 0

        run_in_thread flow_controller, :acquire, 40, adding_1_done
        assert adding_1_done.wait(0.1), "Adding message 1 never unblocked."

        run_in_thread flow_controller, :acquire, 80, adding_2_done
        refute adding_2_done.wait(0.1), "Adding message 2 did not block."

        flow_controller.release 40

        run_in_thread flow_controller, :acquire, 50, adding_3_done

        assert adding_2_done.wait(5), "Adding message 2 never unblocked. test loop: #{i}"
        flow_controller.release 80
        assert adding_3_done.wait(0.1), "Adding message 3 never unblocked. test loop: #{i}"
      end
    end
```

It fails consistently:

```
Adding message 2 never unblocked. test loop: 55
Adding message 2 never unblocked. test loop: 39
Adding message 2 never unblocked. test loop: 18
Adding message 2 never unblocked. test loop: 12
Adding message 2 never unblocked. test loop: 26
```

Now I'll try your proposed solution. Thanks!

Reply:
Updated, PTAL, thanks.